src/snp.rs

```diff
@@ -0,0 +1,265 @@
+        }
+    }
+}
+
+impl Attester for SnpAttester {
+    fn workload_id(&self) -> &String {
+        &self.workload_id
+    }
+
+    fn challenge(&mut self) -> Result<Challenge, AttesterError> {
+        /*
+         * For SNP, we're really only concerned about the nonce in the
+         * Challenge. Therefore, we have no reason to create an SnpChallenge
+         * struct. Use the generic Challenge to send back to the client.
+         */
+        Ok(Challenge {
+            nonce: self.nonce.clone(),
+            extra_params: "".to_string(),
+        })
+    }
+
+    fn attest(
+        &mut self,
+        attestation: &Attestation,
+        measurement: &str,
+    ) -> Result<(), AttesterError> {
+        let snp: SnpAttestation = json::from_str(&attestation.tee_evidence).unwrap();
+        let report: AttestationReport = json::from_str(&snp.report).unwrap();
+        let pkey: &TeePubKey = &attestation.tee_pubkey;
+        if !nonce_pkey_hash(&report, pkey, &self.nonce) {
+            return Err(AttesterError::SnpNoncePubkeyHashInvalid);
+        }
+
+        if measurement.as_bytes() != report.measurement {
+            return Err(AttesterError::SnpMeasurementInvalid);
+        }
+
+        if validate_chain(&report, SnpGeneration::Milan).is_err() {
+            return Err(AttesterError::SnpCertChainInvalid);
+        }
+
+        self.rsa = match rsa_from_tee_pubkey_components(pkey) {
+            Ok(key) => Some(key),
+            Err(_) => return Err(AttesterError::TeePubkeyInvalid),
+        };
+
+        Ok(())
+    }
+
+    fn encrypt_secret(&self, plain_secret: &[u8]) -> Result<Value, AttesterError> {
+        /*
+         * At this point, self.rsa should contain a valid Rsa, and thus it is
+         * safe to unwrap().
+         */
+        let rsa = self.rsa.as_ref().unwrap();
+        let mut encrypted_secret = vec![0; rsa.size() as usize];
+
+        if rsa
+            .public_encrypt(plain_secret, &mut encrypted_secret, Padding::PKCS1)
+            .is_err()
+        {
+            return Err(AttesterError::SnpSecretEncryption);
+        }
+
+        let hex_json = json::json!(hex::encode(encrypted_secret));
+
+        Ok(hex_json)
+    }
+}
+
+fn nonce_pkey_hash(report: &AttestationReport, pkey: &TeePubKey, nonce: &str) -> bool {
+    let hash_expect = {
+        let mut sha = Sha512::new();
+
+        sha.update(nonce.as_bytes());
+        sha.update(pkey.k_mod[..].as_bytes());
+        sha.update(pkey.k_exp[..].as_bytes());
+
+        sha.finish()
+    };
+
+    hash_expect == report.report_data
+}
+
+fn validate_chain(report: &AttestationReport, gen: SnpGeneration) -> Result<(), AttesterError> {
+    let ca = CaChain::get(&gen);
+    let vcek = {
+        let id = hex::encode(&report.chip_id);
+        let url = format!("https://kdsintf.amd.com/vcek/v1/{}/{}?blSPL={:02}&teeSPL={:02}&snpSPL={:02}&ucodeSPL={:02}",
+                gen.to_proper(), id, report.current_tcb.boot_loader, report.current_tcb.tee,
```

Comment:
`gen.to_proper()` is here because the URL for a VCEK is different between Milan and Genoa machines. The question is: how do we get client's to understand which architecture they are running on (Milan or Genoa) and give this information to the attestation server. This may require a change in the TEE config?

Reply:
I think extending TEE config  with a generic field where we can put platform-specific data can be a good idea here.

Reply:
Ok. Will try that out in libkrun and extend the `kbs_types::SnpAttestation` field to include a `String` for the generation.